docs/searching.rst

```diff
@@ -0,0 +1,348 @@
+* Through adding *criteria.*
+
+Text Query Support
+******************
+
+Text queries may be added to a query object by using the query object's
+:func:`where() <cbc_sdk.base.QueryBuilder.where>`, :func:`and_() <cbc_sdk.base.QueryBuilder.and_>`, and
+:func:`or_() <cbc_sdk.base.QueryBuilder.or_>` methods.  The following example sets up a query looking for events
+in which the program ``googleupdate.exe`` accesses the system registry on a device with a specific hostname, IP
+address, and owner::
+
+    # assume the CBCloudAPI object is in the variable "api"
+    from cbc_sdk.endpoint_standard import Event
+    >>> event_query = api.select(Event).where(hostName='Win10').and_(ipAddress='10.0.0.1')
+
+    # further refine the query
+    >>> event_query.and_(applicationName='googleupdate.exe').and_(eventType='REGISTRY_ACCESS')
+    >>> event_query.and_(ownerNameExact='DevRel')
+
+The ``where()`` method supplies the initial query parameters, while ``and_()`` and ``or_()`` add additional query
+parameters. As with other languages, ``and_()`` gets grouped together before ``or_()``.
+
+Parameters may either be supplied as text strings or as keyword assignments::
+
+    >>> from cbc_sdk.platform import Device
+    # the following two queries are equivalent
+    >>> string_query = api.select(Device).where("avStatus:AV_ACTIVE")
+    >>> keyword_query = api.select(Device).where(avStatus="AV_ACTIVE")
+
+However, mixing the two types in a single query is not allowed::
+
+    # this is not allowed
+    >>> from cbc_sdk.platform import Device
+    >>> bogus_query = api.select(Device).where(avStatus="AV_ACTIVE").and_("virtualMachine:true")
+    cbc_sdk.errors.ApiError: Cannot modify a structured query with a raw parameter
+
+Criteria Support
+****************
+
+Criteria are usually added to queries using methods specific to each query.  For example, this query looks for alerts
+with severity 9 or 10 on a machine running macOS 10.14.6::
+
+    >>> from cbc_sdk.platform import Alert
+    >>> alert_query = api.select(Alert)
+
+    # Refine the query with parameters
+    >>> alert_query.where(alert_severity=9).or_(alert_severity=10)
+
+    # Refine the query with criteria
+    >>> alert_query.set_device_os(["MAC"]).set_device_os_versions(["10.14.6"])
+
+This query produces the following JSON block to be passed to a ``POST`` request to the server:
+
+.. code-block:: json
+
+  {
+    "query": "alert_severity:9 OR alert_severity:10",
+    "criteria": {
+      "device_os": ["MAC"],
+      "device_os_version": ["10.14.6"]
+    }
+  }
+
+In newer queries, the various specific methods for setting each individual criterion will be replaced with a single
+method::
+
+    # Refine the query with criteria (new style)
+    >>> alert_query.set_criteria("device_os", ["MAC"]).set_criteria("device_os_version", ["10.14.6"])
+
+.. note::
+
+    The ``set_criteria()`` method is supported with Alerts v7, and will be supported in more classes over time.
```

Comment:
We already support this in a bunch of models as `add_criteria` using the CriteriaBuilderSupportMixin

Reply:
Rephrasing.